docs/rules/no-mutation-props.md

```diff
@@ -0,0 +1,135 @@
+  render: function() {
+    const [firstThing] = this.props.list;
+    firstThing.foo = 'bar';
+    return <div>{firstThing.foo}</div>;
+  }
+});
+
+var Hello = React.createClass({
+  render: function() {
+    delete this.props.foo;
+    return <div>{Object.keys(this.props).length} props</div>;
+  }
+});
+
+var Hello = React.createClass({
+  render: function() {
+    Object.assign(this.props.foo, {bar: 'baz'});
+    return <div>{this.props.foo.bar}</div>;
+  }
+});
+
+var Hello = React.createClass({
+  render: function() {
+    Object.defineProperty(this.props, 'foo');
+    return <div>{this.props.foo}</div>;
+  }
+});
+
+var Hello = React.createClass({
+  render: function() {
+    this.props.foo++;
+    return <div>{this.props.foo}</div>;
+  }
+});
+```
+
+## Rule Options
+
+```js
+...
+"react/no-mutation-props": [<enabled>, {
+  "allowArrayMutations": <boolean>,
+  "allowableArrayMutations": <array<string>>,
+  "allowObjectStatics": <boolean>,
+  "allowableObjectStatics": <array<string>>,
+}]
+...
+```
+
+### `allowArrayMutations`
+
+When `true` the rule ignores [array mutation](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array#Mutator_methods) of props.
+
+You should only use this option if you have a prop that contains an object with methods that conflict with array mutation methods.
+
+The following patterns are considered okay and do not cause warnings:
+
+```jsx
+this.props.list.push(1)
+```
+
+### `allowablArrayMutations`
+
+When `allowArrayMutation: true` the rule specifies specific [array mutations](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array#Mutator_methods) of props to allow. Mutations not listed will remain as warnings.
+
+You must used this option in conjunction with `allowArrayMutation`. You should only use this option if you have a prop that contains an object with methods that conflict with array mutation methods.
+
+### `allowObjectStatics`
+
+When `true` the rule ignores [`Object` methods](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/object#Methods_of_the_Object_constructor) that can mutation props.
+
+You should only use this option if you've overridden the language built-in `Object`.
+
+The following patterns are considered okay and do not cause warnings:
+
+```jsx
+Object.assign(this.props, {foo: true})
+Object.defineProperty(this.props.foo, 'bar')
+Object.defineProperties(this.props.foo, [{foo: {value: 'bar'}}])
+```
+
+### `allowableObjectStatics`
+
+When `allowObjectStatics: true` the rule specifies specific [`Object` methods](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/object#Methods_of_the_Object_constructor) of `Object` to allow. Mutation methods not listed will remain as warnings.
+
+You must used this option in conjunction with `allowObjectStatics`. You should only use this option if you've overridden the language built-in `Object`.
+
+### `allowReflectStatics`
```

Comment:
Instead of Object and Reflect statics, this really needs to just be "statics" - all of them should appear in the same list, and I should be able to define other static methods on other objects (like eslint's `no-restricted-properties` rule) and have it work.

Reply:
I understand and want the same thing, but it's significantly more complex to support arbitrary objects and their statics. It seems to me that that this API, while not ideal, is a good v1. We can add complexity down the road.

Reply:
eslint core handles it with `{ object: 'Object', property: 'assign', message: 'some custom message' }` - I'm not sure why we couldn't handle it the same way (as `no-restricted-properties`)

Reply:
Sure! That would work for `Reflect` and `Object` (though I'm not sure it's a more clear API), but it wouldn't enable lodash, or any other non-built-in as far as I know. Unless there's something I'm missing?

Reply:
`{ property: 'lodash' }` is used in eslint core to refer to the bare identifier.

Reply:
Oh, interesting. So, that doesn't actually guarantee that the object is imported from lodash. e.g. it could throw a false positive on something like `const lodash = {assign: noop}`, and wouldn't catch `import _ from 'lodash'`, nor `import {assign} from 'lodash'`. I can do that, but it feels a little gross. Thoughts?

Reply:
(I get that you could set `_.assign` in the `statics` array, but as I say … it's a bit gross)

Reply:
Yes, but I don't think there's much of a way around that, since adding awareness of imports/requires would be a bit impractical.

Reply:
Okay, fair enough. I'll add that! :)